pyesbulk/__init__.py

```diff
@@ -68,23 +101,58 @@ def put_template(es, name, body):
 
         Failure modes are raised as exceptions.
     """
+    assert name is not None and mapping_name is not None and body is not None
     retry = True
     retry_count = 0
     backoff = 1
     beg, end = time.time(), None
+    try:
+        body_ver = int(body["mappings"][mapping_name]["_meta"]["version"])
+    except KeyError as e:
+        raise Exception(
+            f"Bad template, {name}, given mapping name misssing from template,"
+            f" {mapping_name}: {e!r}"
+        )
     while retry:
+        try:
+            tmpl = es.indices.get_template(name=name)
+        except es_excs.ConnectionError:
+            # We retry all connection errors
+            _sleep_w_backoff(backoff)
+            backoff += 1
+            retry_count += 1
+            continue
+        except es_excs.TransportError as exc:
+            # Only retry on certain 500 errors
+            if exc.status_code != 404:
+                if exc.status_code not in [500, 503, 504]:
+                    raise
+                _sleep_w_backoff(backoff)
+                backoff += 1
+                retry_count += 1
+                continue
+        else:
+            try:
+                tmpl_ver = int(
+                    tmpl[name]["mappings"][mapping_name]["_meta"]["version"]
+                )
+            except KeyError:
+                pass
+            else:
+                if tmpl_ver == body_ver:
+                    break
         try:
             es.indices.put_template(name=name, body=body)
-        except es_excs.ConnectionError as exc:
+        except es_excs.ConnectionError:
             # We retry all connection errors
-            time.sleep(_calc_backoff_sleep(backoff))
+            _sleep_w_backoff(backoff)
             backoff += 1
             retry_count += 1
         except es_excs.TransportError as exc:
             # Only retry on certain 500 errors
             if exc.status_code not in [500, 503, 504]:
                 raise
-            time.sleep(_calc_backoff_sleep(backoff))
+            _sleep_w_backoff(backoff)
             backoff += 1
             retry_count += 1
         else:
```

Comment:
And we don't worry about the "re-try-able" 500 errors here at all? Granted, probably rare, but since we're putting a known document to replace a known name, seems a retry would still be idempotent?

Reply:
With the op-type of "create" this is okay, because we ensure the caller is providing the document IDs, which means we'll get a duplicate back on re-try if we received a 500 error, but somehow the document was also indexed.

For the "update" op-type, this is also okay, because the document will just get updated again, but the expected error if the document ID does not already exist.

For the "delete" op-type, this is also okay, because if it was deleted, we'll just get a 404 back.

For the "index" op-type, if the ID already exists, it will be updated again, and if it doesn't already exist it will be created.

Seems like this is a good comment to add...

Reply:
> And we don't worry about the "re-try-able" 500 errors here at all?

Can you explain more what you mean?  This code will retry all 500, 503, and 504 errors.  Any other 5xx error re-raise the `TransportError` exception.

What do we have to worry about when retrying PUT operations on a template?

Reply:
Ha ha; somehow I completely missed that the check was repeated here. Weird.

Although maybe it'd be better to capture the non-retryable list in a CONSTANT so it's immediately clear we're using the same set in both places.

Reply:
Fixed.